File: Python/RandomPickIndex/RandomPickIndex.py
import random

class Solution(object):

    def __init__(self, nums):
        """
        
        :type nums: List[int]
        :type numsSize: int
        """
        self.nums = nums
        
    # (122ms)
    def pick(self, target):
        """
        :type target: int
        :rtype: int
        """
        ans, count = -1, 0
        for index, num in enumerate(self.nums):
            if num == target:
                if random.randint(0, count) == 0:
                    ans = index
                count += 1
        return ans
        


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.pick(target)

File: Java/RandomPickIndex/RandomPickIndex.java
import java.util.Random;

/**
 * (307ms)
 * 蓄水池算法
 * Created by Phoenix-Z on 2017/6/10.
 */
public class RandomPickIndex {

    private int[] nums;
    private Random random;

    public RandomPickIndex(int[] nums) {
        this.nums = nums;
        random = new Random();
    }

    public int pick(int target) {
        int ans = -1;
        int count = 0;
        for(int i = 0; i < nums.length; i++) {
            if(nums[i] == target) {
                if(random.nextInt(++count) == 0)
                    ans = i;
            }
        }
        return ans;
    }

}
